### ISSUE TITLE
Fix dependency versions

### ISSUE THREAD

**Issue by EECOLOR** [reactjs/redux#1938]
**Do you want to request a *feature* or report a *bug*?**  

I guess if I had to choose it would be a feature.

**What is the current behavior?**

When I run `npm install` dependencies of redux are updated, potentially breaking it.

**What is the expected behavior?**

That a fixed version of redux has fixed dependencies.

**Which versions of Redux, and which browser and OS are affected by this issue? Did this work in previous versions of Redux?**

I don't think so. I think this has been in all versions.

Semantic versioning is great, but it currently not enforced in javascript (I believe Elm is de only language with that feature). This means we need to trust the developer of the library to correctly implement semantic versioning. In the past quite a few libraries have (accidentally) broken something, which is natural, we are all humans.

In order to ensure a stable application that has its dependencies tested we fix the versions of the libraries we use. When we update a dependency we test the application to see if nothing was broken. When a library however has transitive dependencies (the ones in the `dependencies` key of `package.json`) that are marked with a `^` we are in trouble.

Would you be willing to remove the `^` in the dependencies of your `package.json`?

**Comment by SimenB** [comment]
Updating package.json doesn't really help, as your dependencies probably still uses `~` or `^`.  Shrinkwrapping would do it, though (it recursively locks all dependencies).
That said, I prefer that to be the responsibility of the consumer, not a library. Give the consumer the flexibility to decide themselves, don't enforce one way of doing it.

**Comment by EECOLOR** [comment]
We had a problem where shrinkwrap did not work for us. I can't remember the details and might look into it again.

However I am still convinced that as a library author you should make a best effort to deliver a working version of your library. If you specify your dependencies using a `^` you can not guarantee that you have tested the version of library in combination with that set of dependencies.

This an example of where encountered this exact problem today: https://github.com/yiminghe/koa-webpack-dev-middleware/issues/10

**Comment by SimenB** [comment]
I don't want locked dependencies, I like to have latest bug fixes without waiting for a new release of a dependent. This would be impossible with the change you suggest (unless it's a dependency of a dependency).

But if I want dependencies locked down (which I might for certain projects at work, perhaps), I use shrinkwrap, which also locks the dependencies of my dependencies, and not just the top level.

So not only does shrinkwrap allow me to still get updated deps, it also also you to actually lock down all dependencies, instead of just the top level. More flexibility, and it works better.

**Comment by EECOLOR** [comment]
> I don't want locked dependencies, I like to have latest bug fixes without waiting for a new release of a dependent.

Even when those bug fixes break your library?

**Comment by SimenB** [comment]
I haven't experienced it yet.
For me, it boils down to: applications might want to lock stuff down, libraries shouldn't.

Either way, I like to have the choice, not have it taken away from me 😄

**Comment by EECOLOR** [comment]
Fair enough, I understand your position.

The only small thing that bothers me is that a `^` communicates: this version of library works until the next major version of the library. This of course is not a very reliable statement. It's quite possible that somewhere between `4.2.1` and `5.0.0` of `lodash` a breaking change will creep in.

Your suggestion of using [Greenkeeper](https://greenkeeper.io/) sounds like a solution for the 'guarantee' given by stating your dependency with `^`.

> I haven't experienced it yet.

This past year I have encountered problems that were caused by `^` dependencies of libraries multiple times.

In any case, it seems we both sit on another side of the trade-offs and with Greenkeeper as a tool I can stand behind your reasoning. Thank you for the discussion.
